_posts/2019-08-23-unet/unet.Rmd

```diff
@@ -0,0 +1,437 @@
+
+So what, technically, is image segmentation, and how can we train a neural network to do it?
+
+## Image segmentation in a nutshell
+
+Say we have an image with a bunch of cats in it. In _classification_, the question is "what's that?", and the answer we want to hear is: "cat". In _object detection_, we again ask "what's that", but now that "what" is implicitly plural, and we expect an answer like "there's a cat, a cat and a cat, and they're here, here, and here" (imagine the network pointing, by means of drawing _bounding boxes_, i.e., rectangles around the detected objects). In _segmentation_, we want more: We want the whole image covered by "boxes" -- which aren't boxes anymore, but unions of pixel-size "boxlets" -- or put differently: __We want the network to label every single pixel in the image.__
+
+Here's an example from the paper we're going to talk about in a second. On the left is the input image (HeLa cells), next up is the ground truth, and third is the learned segmentation mask.
+
+```{r, eval=TRUE, echo=FALSE, layout="l-body-outset", fig.cap = "Example segmentation from Ronneberger et al. 2015."}
+knitr::include_graphics("images/mask.png")
+```
+
+Technically, a distinction is made between _class segmentation_ and _instance segmentation_. In class segmentation, referring to the "bunch of cats" example, there are two possible labels: Every pixel is either "cat" or "not cat". Instance segmentation is more difficult: Here every cat gets their own label. (As an aside, why should that be more difficult? Presupposing human-like cognition, it wouldn't be -- if I have the concept of a cat, instead of just "cattiness", I "see" there are two cats, not one. But depending on what a specific neural network relies on most -- texture, color, isolated parts -- those tasks may differ a lot in difficulty.)
+
+The network architecture used in this post is adequate for _class segmentation_ tasks and should be applicable to a vast number of practical, scientific as well as non-scientific applications. Speaking of network architecture, how should it look?
+
+## Introducing U-Net
+
+Given their success in image classification, can't we just use a classic architecture like _Inception V[n]_, _ResNet_, _ResNext_ ... , whatever? The problem is, our task at hand -- labeling every pixel -- does not fit so well with the classic idea of a CNN. With convnets, the idea is to apply successive layers of convolution and pooling to build up feature maps of decreasing granularity, to finally arrive at an abstract level where we just say: "yep, a cat". The other side of the medail is, we lose detail information: To the final classification, it does not matter whether the five pixels in the top-left area are black or white.
```

Comment:
I don't think you've spelled out CNN yet in this piece (maybe give it a hover-able "convolutional neural networks"?)

Reply:
> medail

I'm not sure what this is. Are you saying something to the effect of _the downside_?

Reply:
oh yeah right ... actually was a typo, should have been "medal", but looks like it's unidiomatic anyway...